substrate/client/rpc-servers/src/middleware/chain_head.rs

```diff
@@ -0,0 +1,237 @@
+#[derive(Clone)]
+pub struct ChainHeadLayer {
+	connection_data: Arc<Mutex<ConnectionData>>,
+}
+
+impl ChainHeadLayer {
+	/// Create a new [`ChainHeadLayer`].
+	pub fn new(connection_data: Arc<Mutex<ConnectionData>>) -> Self {
+		Self { connection_data }
+	}
+}
+
+impl<S> tower::Layer<S> for ChainHeadLayer {
+	type Service = ChainHeadMiddleware<S>;
+
+	fn layer(&self, inner: S) -> Self::Service {
+		ChainHeadMiddleware::new(inner, self.connection_data.clone())
+	}
+}
+
+/// Chain head middleware.
+#[derive(Clone)]
+pub struct ChainHeadMiddleware<S> {
+	service: S,
+	connection_data: Arc<Mutex<ConnectionData>>,
+}
+
+impl<S> ChainHeadMiddleware<S> {
+	/// Create a new chain head middleware.
+	pub fn new(service: S, connection_data: Arc<Mutex<ConnectionData>>) -> ChainHeadMiddleware<S> {
+		ChainHeadMiddleware { service, connection_data }
+	}
+}
+
+impl<'a, S> RpcServiceT<'a> for ChainHeadMiddleware<S>
+where
+	S: Send + Sync + RpcServiceT<'a>,
+{
+	type Future = ResponseFuture<S::Future>;
+
+	fn call(&self, req: Request<'a>) -> Self::Future {
+		const CHAIN_HEAD_FOLLOW: &str = "chainHead_unstable_follow";
+		const CHAIN_HEAD_CALL_METHODS: [&str; 8] = [
+			"chainHead_unstable_body",
+			"chainHead_unstable_header",
+			"chainHead_unstable_call",
+			"chainHead_unstable_unpin",
+			"chainHead_unstable_continue",
+			"chainHead_unstable_storage",
+			"chainHead_unstable_stopOperation",
+			"chainHead_unstable_unfollow",
+		];
+
+		let method_name = req.method_name();
+
+		// Intercept the subscription ID returned by the `chainHead_follow` method.
+		if method_name == CHAIN_HEAD_FOLLOW {
+			return ResponseFuture::Register {
+				fut: self.service.call(req.clone()),
+				connection_data: self.connection_data.clone(),
+			}
+		}
+
+		// Ensure the subscription ID of those methods corresponds to a subscription ID
+		// of this connection.
+		if CHAIN_HEAD_CALL_METHODS.contains(&method_name) {
+			let params = req.params();
+			let follow_subscription = get_subscription_id(params);
+
+			if let Some(follow_subscription) = follow_subscription {
+				if !self.connection_data.lock().subscriptions.contains(&follow_subscription) {
+					return ResponseFuture::Ready {
+						response: Some(MethodResponse::error(
+							req.id(),
+							jsonrpsee::types::error::ErrorObject::owned(
+								-32602,
+								"Invalid subscription ID",
+								None::<()>,
+							),
+						)),
+					};
+				}
+			}
+		}
+
+		ResponseFuture::Forward { fut: self.service.call(req.clone()) }
+	}
+}
+
+/// Extract the subscription ID from the provided parameters.
+///
+/// We make the assumption that all `chainHead` methods are given the
+/// subscription ID as a first parameter.
+///
+/// This method handles positional and named `camelCase` parameters.
+fn get_subscription_id<'a>(params: Params<'a>) -> Option<String> {
+	// Support positional parameters.
+	if let Ok(follow_subscription) = params.sequence().next::<String>() {
```

Comment:
you can call `params.is_object()` here to do either parse it as a sequence or map.

however, you use `params.one::<String>() instead sequence here since you only care about one item